File: Client/Gamify.Client.Net/Gamify.Client.Net/Managers/GameConnectionManager.cs
using Gamify.Client.Net.Client;
using Gamify.Client.Net.Services;
using Gamify.Contracts.Notifications;
using Gamify.Contracts.Requests;
using System;

namespace Gamify.Client.Net.Managers
{
    public class GameConnectionManager
    {
        private readonly IGameService<PlayerConnectRequestObject, PlayerConnectedNotificationObject> playerConnectService;
        private readonly IGameService<PlayerDisconnectRequestObject, PlayerDisconnectedNotificationObject> playerDisconnectService;
        private readonly IGameService<GetConnectedPlayersRequestObject, SendConnectedPlayersNotificationObject> connectedPlayersService;

        public event EventHandler<GameNotificationEventArgs<PlayerConnectedNotificationObject>> PlayerConnectedNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<PlayerDisconnectedNotificationObject>> PlayerDisconnectedNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<SendConnectedPlayersNotificationObject>> ConnectedPlayersNotificationReceived;

        public GameConnectionManager(string playerName, IGameClientFactory clientFactory)
        {
            var gameClient = clientFactory.GetGameClient(playerName);

            this.playerConnectService = new GameService<PlayerConnectRequestObject, PlayerConnectedNotificationObject>(GameRequestType.PlayerConnect, GameNotificationType.PlayerConnected, gameClient);
            this.playerDisconnectService = new GameService<PlayerDisconnectRequestObject, PlayerDisconnectedNotificationObject>(GameRequestType.PlayerDisconnect, GameNotificationType.PlayerDisconnected, gameClient);
            this.connectedPlayersService = new GameService<GetConnectedPlayersRequestObject, SendConnectedPlayersNotificationObject>(GameRequestType.GetConnectedPlayers, GameNotificationType.SendConnectedPlayers, gameClient);

            this.playerConnectService.NotificationReceived += (sender, args) =>
            {
                this.NotifyPlayerConnected(args);
            };

            this.playerDisconnectService.NotificationReceived += (sender, args) =>
            {
                this.NotifyPlayerDisconnected(args);
            };

            this.connectedPlayersService.NotificationReceived += (sender, args) =>
            {
                this.NotifyConnectedPlayers(args);
            };
        }

        public void ConnectPlayer(PlayerConnectRequestObject playerConnectRequest)
        {
            this.playerConnectService.Send(playerConnectRequest);
        }

        public void DisconnectPlayer(PlayerDisconnectRequestObject playerDisconnectRequest)
        {
            this.playerDisconnectService.Send(playerDisconnectRequest);
        }

        private void NotifyPlayerConnected(GameNotificationEventArgs<PlayerConnectedNotificationObject> args)
        {
            if (this.PlayerConnectedNotificationReceived != null)
            {
                this.PlayerConnectedNotificationReceived(this, args);
            }
        }

        private void NotifyPlayerDisconnected(GameNotificationEventArgs<PlayerDisconnectedNotificationObject> args)
        {
            if (this.PlayerDisconnectedNotificationReceived != null)
            {
                this.PlayerDisconnectedNotificationReceived(this, args);
            }
        }

        private void NotifyConnectedPlayers(GameNotificationEventArgs<SendConnectedPlayersNotificationObject> args)
        {
            if (this.ConnectedPlayersNotificationReceived != null)
            {
                this.ConnectedPlayersNotificationReceived(this, args);
            }
        }
    }
}

File: Java/gamify-java-client/src/main/java/ar/com/mdf/gamify/managers/GameConnectionManager.java
package ar.com.mdf.gamify.managers;

import java.net.URISyntaxException;

import ar.com.mdf.gamify.client.IGamifyClient;
import ar.com.mdf.gamify.enums.NotificationType;
import ar.com.mdf.gamify.enums.RequestType;
import ar.com.mdf.gamify.exceptions.NotConnectedException;
import ar.com.mdf.gamify.factory.IClientFactory;
import ar.com.mdf.gamify.objects.notification.ConnectionStatePlayerNotificationObject;
import ar.com.mdf.gamify.objects.notification.SendConnectedPlayersNotificationObject;
import ar.com.mdf.gamify.objects.request.GetConnectedPlayersRequestObject;
import ar.com.mdf.gamify.objects.request.PlayerNameRequestObject;
import ar.com.mdf.gamify.observers.IGameObserver;
import ar.com.mdf.gamify.services.GameService;
import ar.com.mdf.gamify.services.IGameService;

public class GameConnectionManager {
	private IGameService<PlayerNameRequestObject, ConnectionStatePlayerNotificationObject> playerConnectService;
	private IGameService<PlayerNameRequestObject, ConnectionStatePlayerNotificationObject> playerDisconnectService;
	private IGameService<GetConnectedPlayersRequestObject, SendConnectedPlayersNotificationObject> connectedPlayersService;
	private IGamifyClient client;
	
	public GameConnectionManager(IClientFactory clientFactory) throws NotConnectedException {
		try {
			this.client = clientFactory.getClient();
		} catch (URISyntaxException e) {
			throw new NotConnectedException(e);
		}
		this.playerConnectService = new GameService<PlayerNameRequestObject, ConnectionStatePlayerNotificationObject>(RequestType.PlayerConnect, NotificationType.PlayerConnected, this.client, ConnectionStatePlayerNotificationObject.class);
		this.playerDisconnectService = new GameService<PlayerNameRequestObject, ConnectionStatePlayerNotificationObject>(RequestType.PlayerDisconnect, NotificationType.PlayerDisconnected, this.client, ConnectionStatePlayerNotificationObject.class);
		this.connectedPlayersService = new GameService<GetConnectedPlayersRequestObject, SendConnectedPlayersNotificationObject>(RequestType.GetConnectedPlayers, NotificationType.SendConnectedPlayers, this.client, SendConnectedPlayersNotificationObject.class);
	}
	
	public void suscribePlayerConnect(IGameObserver<ConnectionStatePlayerNotificationObject> observer){
		this.playerConnectService.suscribe(observer);
	}
	public void suscribePlayerDisconnect(IGameObserver<ConnectionStatePlayerNotificationObject> observer){
		this.playerDisconnectService.suscribe(observer);
	}
	public void suscribeConnectedPlayers(IGameObserver<SendConnectedPlayersNotificationObject> observer){
		this.connectedPlayersService.suscribe(observer);
	}

	public void connect(PlayerNameRequestObject connectRequest) {
		this.playerConnectService.send(connectRequest);
	}
	public void disconnect(PlayerNameRequestObject disconnectRequest) {
		this.playerDisconnectService.send(disconnectRequest);
	}
	public void rejectGame(GetConnectedPlayersRequestObject connectedPlayersRequest) {
		this.connectedPlayersService.send(connectedPlayersRequest);
	}
}
